pkg/autoscaler/aggregation/bucketing.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2019 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aggregation
+
+import (
+	"sync"
+	"time"
+)
+
+// TimedFloat64Buckets keeps buckets that have been collected at a certain time.
+type TimedFloat64Buckets struct {
+	bucketsMutex sync.RWMutex
+	buckets      map[time.Time]Float64Bucket
+
+	granularity time.Duration
+}
+
+// NewTimedFloat64Buckets generates a new TimedFloat64Buckets with the given
+// granularity.
+func NewTimedFloat64Buckets(granularity time.Duration) *TimedFloat64Buckets {
+	return &TimedFloat64Buckets{
+		buckets:     make(map[time.Time]Float64Bucket),
+		granularity: granularity,
+	}
+}
+
+// Record adds a value with an associated time to the correct bucket.
+func (t *TimedFloat64Buckets) Record(time time.Time, name string, value float64) {
+	t.bucketsMutex.Lock()
+	defer t.bucketsMutex.Unlock()
+
+	bucketKey := time.Truncate(t.granularity)
+	bucket, ok := t.buckets[bucketKey]
```

Comment:
at this point this is un-necessary
`return t.buckets[key].Record()` will work.

Reply:
Tried that, doesn't work. `Float64Bucket` is still a map and the default value for that is nil.

Reply:
Yeah, my wrong
🤦‍♂ 

/lgtm
/approve